### PR TITLE
Feature more dynamic configuration update

### PR DESCRIPTION
**Pull request by khimstedt**
# Short Preface

There were many recent changes in the mistral.yaml machine file in the branch "develop" on which this feature branch is based on, and as a consequence the build step for awicmcr-CMIP6 was broken. As a workaround the settings from a previous version of mistral.yaml are used here to succesfully build awicmcr-CMIP6 on Mistral.

Additionally affected files:

... /esm_tools/configs/esm_master/components/echam.yaml
... /esm_tools/configs/esm_master/components/oasis3-mct.yaml

A step-by-step approach during the merge is probably the best way to find out what change has initially broken the ECHAM Concurrent Radiation build process with the ESM-Tools and to bring all different models and clusters under one roof.

# Feature Description

With this feature, typical configuration changes can be made by users more dynamically and at a central location (e.g. selecting the MPI environment, building a debug version of a program, selecting the hyperthreading mode or specifying launcher flags).
For this purpose, the esm_master or esm_runscripts command is given an additional configuration yaml file as argument (--modify-config), for example usermods.yaml (if needed with relative or absolute path specification).

Example content of usermods.yaml:

```yaml
build_and_run_modifications:
  machine:
    chooseable_settings:
      computer.useMPI: intel18_bullxmpi

build_only_modifications:
  machine:
    environment_settings:
     computer.export_vars.ESM_DEBUG_ECHAM: 'TRUE'
     computer.export_vars.ESM_DEBUG_FESOM: 'TRUE'

run_only_modifications:
  machine:
    chooseable_settings:
#     general.use_hyperthreading: '0'

  batch_system:
    direct_settings:
#     computer.launcher_flags: "-l --kill-on-bad-exit=1 --cpu_bind=cores --distribution=cyclic:cyclic"

# kh 22.07.20 can be used as a "low level" alternative to general.use_hyperthreading: '0' (see above)
      computer.hyperthreading_flag: "--ntasks-per-core=1"
```

There are three main sections in usermods.yaml:
a)
build_and_run_modifications refers to changes that are relevant for the build step as well as for the run step (e.g. the selection of the MPI environment)

b)
build_only_modifications refers to changes that are relevant for the build step (such as building a debug or release version)

c)
run_only_modifications refers to changes that are relevant for the run step (such as selecting hyperthreading or launcher flags)

The structure of a leaf-level entry that contains the path to the appropriate dictionary or list object is more or less self-explanatory (e.g. computer.useMPI: intel18_bullxmpi).
Basically, these modifications could also be made in the existing configuration files, which, apart from the runscripts, are rather static in nature (i.e. typically not necessarily modified by the user). The intermediate level items (machine, batch_system) in usermods.yaml - redundant in principle - give an indication of the configuration section in which an entry is originally set.

In addition, --ignore-errors is another new argument for esm_master or esm_runscripts to avoid an early abort in case of an error. This can sometimes be helpful especially during the development of the ESM-Tools.

### CODE REVIEW

**Comment by dbarbi** [comment]
ok, i merged esm_runscripts with the same pull request, but i don't see anything related to more dynamic configurations in this branch . this is all awicmcr and echamcr, right?

**Comment by dbarbi** [comment]
already merged into develop
